metadata-ingestion/src/datahub/ingestion/source/tableau.py

```diff
@@ -534,49 +566,93 @@ def __init__(
         ctx: PipelineContext,
     ):
         super().__init__(config, ctx)
-
         self.config: TableauConfig = config
         self.report: TableauSourceReport = TableauSourceReport()
         self.server: Optional[Server] = None
-        self.database_tables: Dict[str, DatabaseTable] = {}
-        self.tableau_stat_registry: Dict[str, UsageStat] = {}
-        self.tableau_project_registry: Dict[str, TableauProject] = {}
-        self.workbook_project_map: Dict[str, str] = {}
-        self.datasource_project_map: Dict[str, str] = {}
-
-        # This map keeps track of the database server connection hostnames.
-        self.database_server_hostname_map: Dict[str, str] = {}
-        # This list keeps track of sheets in workbooks so that we retrieve those
-        # when emitting sheets.
-        self.sheet_ids: List[str] = []
-        # This list keeps track of dashboards in workbooks so that we retrieve those
-        # when emitting dashboards.
-        self.dashboard_ids: List[str] = []
-        # This list keeps track of embedded datasources in workbooks so that we retrieve those
-        # when emitting embedded data sources.
-        self.embedded_datasource_ids_being_used: List[str] = []
-        # This list keeps track of datasource being actively used by workbooks so that we only retrieve those
-        # when emitting published data sources.
-        self.datasource_ids_being_used: List[str] = []
-        # This list keeps track of datasource being actively used by workbooks so that we only retrieve those
-        # when emitting custom SQL data sources.
-        self.custom_sql_ids_being_used: List[str] = []
+        self._authenticate(self.config.site)
 
-        self._authenticate()
+    def _authenticate(self, site_content_url: str) -> None:
```

Comment:
The refactoring you proposed (https://github.com/datahub-project/datahub/pull/10498#discussion_r1618146575) was quite straight forward but there was one tricky thing. The `_authenticate` method would've been used in both the `TableauSource` and the new `TableauSiteSource`. We need it in the `TableauSource` to get all the sites and in the new `TableauSiteSource` to re-authenticate in this special case here: https://github.com/datahub-project/datahub/blob/74a543db3fe6773443a7d63b2e45598e06b79286/metadata-ingestion/src/datahub/ingestion/source/tableau.py#L848
Since `_authenticate()` replaces the `server` variable, we can not pass it from `TableauSource` to `TableauSiteSource`.

What I did now is introduce a new method `_re_authenticate()` that is only used in `TableauSiteSource` to do the re-authentication in the mentioned special case. This method now doesn't replace the complete `server` but just signs in again with the current credentials. With this, we can use the same server in both classes and just pass it down to the `TableauSiteSource`.

Only thing I'm not fully sure yet is this part here: https://github.com/datahub-project/datahub/pull/10498/files#diff-f618f3bf4fb8b6cd799b3375ab22051094345681942a27895f8a7d56f3c36951R619
We're using the pager on `server.sites` in the for-loop and I wonder if re-authenticating has an impact on this. Btw, this would've been the same without the refactoring. I did test this and manually triggered a re-authenticate in the middle of a site ingestion and it seemed to work fine for me. Just wanted to point that out. Let me know what you think.

Reply:
One thing we could do is fully materialize the list of sites at the beginning - e.g. do `for site in list(TSC.Pager(self.server.sites):`

That way, even if there's pagination, we will make all the API calls at the start, and don't need to worry about calling reauthenticate from the TableauSource

The fully correct way to do this would be creating a TableauConnection object that wraps the tableau server client and automatically reauthenticates when necessary. However, I don't think we need to do that refactoring here

Reply:
Awesome, thanks for the hint. I'll do that then. 👍